### PR TITLE
[Bifrost] Basic BifrostAdmin interface

### PR DESCRIPTION
**Pull request by AhmedSoliman**
[Bifrost] Basic BifrostAdmin interface

It now hosts `trim` and introduces the ability to seal and extend the chain with a pre-determined loglet params

---
[//]: # (BEGIN SAPLING FOOTER)
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/restatedev/restate/pull/1753).
* #1757
* __->__ #1753
* #1747
* #1744
* #1743
* #1742

### CODE REVIEW

File: crates/bifrost/src/bifrost_admin.rs [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+            metadata_writer,
+            metadata_store_client,
+        }
+    }
+    /// Trim the log prefix up to and including the `trim_point`.
+    /// Set `trim_point` to the value returned from `find_tail()` or `Lsn::MAX` to
+    /// trim all records of the log.
+    ///
+    /// Note that trim does not promise that the log will be trimmed immediately and atomically,
+    /// but they are promise to trim prefixes, earlier segments in the log will be trimmed
```
Comment (tillrohrmann) [inline crates/bifrost/src/bifrost_admin.rs:66]: they seems out of place here.

File: crates/bifrost/src/read_stream.rs [hunk 2]
```diff
@@ -869,49 +882,24 @@ mod tests {
...
+            // seal the loglet and extend with an in-memory one
+            let new_segment_params = new_single_node_loglet_params(ProviderKind::InMemory);
+            bifrost_admin
+                .seal_and_extend_chain(
+                    LOG_ID,
+                    None,
+                    Version::MIN,
+                    ProviderKind::InMemory,
+                    new_segment_params,
+                )
```
Comment (tillrohrmann) [inline crates/bifrost/src/read_stream.rs:894]: Should we assert that sealing has actually happened? Now it seems that the test would also pass if `seal_and_extend_chain` would be a no-op.

File: crates/core/src/metadata_store/mod.rs [hunk 3]
```diff
@@ -167,6 +173,8 @@ impl MetadataStoreClient {
             .await
     }
 
+    /// Deletes the key-value pair for the given key following the provided precondition. If the
+    /// precondition is not met, then the operation returns a [`WriteError::PreconditionViolation`].
```
Comment (tillrohrmann) [inline crates/core/src/metadata_store/mod.rs:177]: Thanks for adding these Rustdocs 🙏

**Review by tillrohrmann — APPROVED** [review]
Nice work @AhmedSoliman. LGTM. +1 for merging :-)

File: crates/bifrost/src/bifrost_admin.rs (hunk 1 above)
Comment (AhmedSoliman) [reply crates/bifrost/src/bifrost_admin.rs:66]: Not sure what you mean.

File: crates/bifrost/src/read_stream.rs (hunk 2 above)
Comment (AhmedSoliman) [reply crates/bifrost/src/read_stream.rs:894]: Good call. Let me do that.

File: crates/bifrost/src/bifrost_admin.rs (hunk 1 above)
Comment (AhmedSoliman) [reply crates/bifrost/src/bifrost_admin.rs]: Removed the paragraph.

**Pull request merged**